### PR TITLE
Chapter 12: Password reset

### PR DESCRIPTION
**Pull request by vantri0964**
Password reset

### CODE REVIEW

**Comment by vantri0964** [comment]
Screenshot from 2019-05-14 07-44-01

**Comment by vantri0964** [comment]
ready

File: app/helpers/account_activations_helper.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
+end
```
Comment (VuThiTranVan) [inline app/helpers/account_activations_helper.rb]: xóa file

File: app/helpers/password_resets_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
+end
```
Comment (VuThiTranVan) [inline app/helpers/password_resets_helper.rb]: xóa file

File: app/mailers/user_mailer.rb [hunk 3]
```diff
@@ -1,11 +1,11 @@
 class UserMailer < ApplicationMailer
   def account_activation user
     @user = user
-    mail to: user.email
+    mail to: user.email, subject: "Account activation"
```
Comment (VuThiTranVan) [inline app/mailers/user_mailer.rb]: I18n

File: app/models/user.rb [hunk 4]
```diff
@@ -52,6 +52,20 @@ def send_activation_email
...
   end
 
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_attribute(:reset_digest, User.digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
```
Comment (VuThiTranVan) [inline app/models/user.rb]: Sử dụng method update 2 field cùng 1 lúc

File: app/models/user.rb [hunk 5]
```diff
@@ -52,6 +52,20 @@ def send_activation_email
...
+  end
+
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
```
Comment (VuThiTranVan) [inline app/models/user.rb]: `2` khai báo vào setting

File: app/views/user_mailer/password_reset.html.erb [hunk 6]
```diff
@@ -1,5 +1,14 @@
...
-
+<h1>
+  <%= t ".title" %>
+</h1>
 <p>
-  <%= @greeting %>, find me in app/views/user_mailer/password_reset.html.erb
+  <% t ".to_reset" %>
+</p>
+<%= link_to t(".link_reset"), edit_password_reset_url(@user.reset_token,
+                                                      email: @user.email) %>
```
Comment (VuThiTranVan) [inline app/views/user_mailer/password_reset.html.erb]: Indent

File: test/controllers/account_activations_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class AccountActivationsControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
```
Comment (VuThiTranVan) [inline test/controllers/account_activations_controller_test.rb]: xóa file

**Comment by vantri0964** [comment]
ready

**Comment by VuThiTranVan** [comment]
merged

**Pull request merged**